proton/cert_pinning.py

```diff
@@ -0,0 +1,98 @@
+import base64
+import hashlib
+import json
+import os
+
+import requests
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.connectionpool import HTTPSConnectionPool
+from requests.packages.urllib3.poolmanager import PoolManager
+
+crypto = requests.packages.urllib3.contrib.pyopenssl.OpenSSL.crypto
+
+class InspectedHTTPSConnectionPool(HTTPSConnectionPool):
+    """Custom HTTPSConnectionPool that verifies the certificate for each connection"""
+    
+    def _validate_conn(self, conn):
+        r = super(InspectedHTTPSConnectionPool, self)._validate_conn(conn)
+ 
+        sock = conn.sock
+        sock_connection = sock.connection
+
+        if self.is_session_secure(sock_connection.get_peer_cert_chain()[0], conn):
+            return r
+
+
+    def is_session_secure(self, cert, conn):
+        """Checks if connection is secure"""
+        
+        cert_hash = self.extract_hash(cert)
+
+        if not self.validate_hash(self.host, cert_hash):
+            # Also generate a report
+            conn.close()
+            raise Exception("Server hash does not match local hash")
+        
+        return True
+
+
+    def validate_hash(self, url, cert_hash):
+        """Validates the hash agains a known list of hashes/pins"""
+        folder_path = os.path.dirname(os.path.abspath(__file__))
+
+        with open(os.path.join(folder_path, "pins.json"), "r") as f:
+            pin_list = json.load(f)
+
+        for domain_info in pin_list["domain_list"]:
+            try:
+                if domain_info["url"] in url and cert_hash in domain_info["hashes"]:
+                    return True
+            except Exception as e:
```

Comment:
What exception can be raised here? In general in a library you should not capture exceptions (unless they are coming from some dependency and you are going to re-launch them or similar).

Reply:
I guess this can be refactored, good point.